### DISCUSSION TITLE
How to get next auto-increment id with sqlalchemy

### DISCUSSION THREAD

**Discussion by ofekisr** [sqlalchemy/sqlalchemy#7629] [Usage Questions]
I develop a flask app with Sqlalchemy (1.3) so the database can be either of the supported dialects.

I'm looking for a way to fetch the next auto-increment id of a specific table (or specific mapper) so I can use it for setting the id inside a model object before I insert it (for test purposes).

I prefer not fetching the max_id of a table since there is a chance that the object who already got the max id was deleted so after fetching the max_id I would get not the next increment id.

**Comment by zzzeek** [comment]
pre-fetching the "next id" is database dependent.  only if you are using a database with SEQUENCE support would this be possible without inventing a custom id generation approach.   for MySQL, SQLite, or any use of IDENTITY, the next primary key is not available until the INSERT is invoked.

I think the best approach here is to rethink this requirement, or if this is only for test purposes, then just manufacture integer identifiers locally and include them within INSERT statements - for an ORM mapped object you can assign it directly.   or, reorganize the test so that it's not dependent on knowing the actual value of the identifier until after the INSERT has proceeded, and at that stage extract the identifier from the newly inserted object after the fact.
